kratos/processes/apply_periodic_boundary_condition_process.cpp

```diff
@@ -0,0 +1,357 @@
+    {
+        const VariableType r_var = KratosComponents<VariableType>::Get(rVarName);
+
+        IndexType master_index = 0;
+        for (auto& master_node : rHostedGeometry)
+        {
+            int current_num_constraint = mrMasterModelPart.NumberOfMasterSlaveConstraints();
+
+            const double master_weight = rWeights(master_index);
+            #pragma omp critical
+            {
+                mrMasterModelPart.CreateNewMasterSlaveConstraint("LinearMasterSlaveConstraint", current_num_constraint++, master_node, r_var, rSalveNode, r_var, master_weight, 0.0);
+            }
+            master_index++;
+        }
+    }
+
+    void ApplyPeriodicConditionProcess::CalculateTransformationMatrix()
+    {
+        if (mType == "translation"){
+            CalculateTranslationMatrix(-1*mModulus, mTransformationMatrix);
+            CalculateTranslationMatrix(0.0, mTransformationMatrixVariable);
+        }
+        else if (mType == "rotation"){
+            CalculateRotationMatrix(-1*mTheta, mTransformationMatrix);
+            CalculateRotationMatrix(mTheta, mTransformationMatrixVariable);
+        }
+    }
+
+    void ApplyPeriodicConditionProcess::CalculateTranslationMatrix(const double Modulus, MatrixType& rMatrix)
+    {
+        rMatrix(0,0) = 1.0;
+        rMatrix(0,1) = 0.0;
+        rMatrix(0,2) = 0.0;
+        rMatrix(0,3) = Modulus *mDirOfTranslation[0];
+        rMatrix(1,0) = 0.0;
+        rMatrix(1,1) = 1.0;
+        rMatrix(1,2) = 0.0;
+        rMatrix(1,3) = Modulus *mDirOfTranslation[1];
+        rMatrix(2,0) = 0.0;
+        rMatrix(2,1) = 0.0;
+        rMatrix(2,2) = 1.0;
+        rMatrix(2,3) = Modulus *mDirOfTranslation[2];
+        rMatrix(3,0) = 0.0;
+        rMatrix(3,1) = 0.0;
+        rMatrix(3,2) = 0.0;
+        rMatrix(3,3) = 1.0;
+    }
+
+    void ApplyPeriodicConditionProcess::CalculateRotationMatrix(const double Theta, MatrixType& rMatrix )
+    {
+        std::vector<double> U(3); // normalized axis of rotation
+        // normalizing the axis of rotation
+        double norm = 0.0;
+        for (unsigned int d = 0; d < 3; ++d)
+            norm += mAxisOfRotationVector[d] * mAxisOfRotationVector[d];
+        norm = sqrt(norm);
+        for (unsigned int d = 0; d < 3; ++d)
+            U[d] = mAxisOfRotationVector[d] / norm;
+
+        // Constructing the transformation matrix
+        const double x1 = mCenterOfRotation[0];
+        const double y1 = mCenterOfRotation[1];
+        const double z1 = mCenterOfRotation[2];
+
+        const double a = U[0];
+        const double b = U[1];
+        const double c = U[2];
+
+        const double t2 = std::cos(Theta);
+        const double t3 = std::sin(Theta);
+        const double t4 = a * a;
+        const double t5 = b * b;
+        const double t6 = c * c;
+        const double t7 = a * b;
+        const double t8 = t5 + t6;
+        double t9 = 1.0 / t8;
+        if(std::isnan(t9))
```

Comment:
how can t9 be nan? (also take care since this test is very very fragile)

Reply:
in that case would checking t8 for being less than epsilon better.

Reply:
yes checking 
        std::abs(t8) < small epsilon

would be definitely more robust.
my point here is that if t8=0 than t9=inf, not nan!